streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -412,29 +410,26 @@ protected void flushState() {
      */
     private void commitOffsets(final boolean startNewTransaction) {
         try {
-            if (commitOffsetNeeded) {
-                log.trace("Committing offsets");
-                final Map<TopicPartition, OffsetAndMetadata> consumedOffsetsAndMetadata = new HashMap<>(consumedOffsets.size());
-                for (final Map.Entry<TopicPartition, Long> entry : consumedOffsets.entrySet()) {
-                    final TopicPartition partition = entry.getKey();
-                    final long offset = entry.getValue() + 1;
-                    consumedOffsetsAndMetadata.put(partition, new OffsetAndMetadata(offset));
-                    stateMgr.putOffsetLimit(partition, offset);
-                }
-
-                if (eosEnabled) {
-                    producer.sendOffsetsToTransaction(consumedOffsetsAndMetadata, applicationId);
-                    producer.commitTransaction();
-                    transactionInFlight = false;
-                    if (startNewTransaction) {
-                        producer.beginTransaction();
-                        transactionInFlight = true;
-                    }
-                } else {
-                    consumer.commitSync(consumedOffsetsAndMetadata);
+            log.trace("Committing offsets");
```

Comment:
This is not a feedback: as we are changing the main loop of StreamThread, we may need to carefully benchmark if this change along with the main loop changes will have unexpected performance penalty: with low traffic input stream, we are effectively sending sync commit requests more frequently. cc @mjsax 

If it does become a problem for performance, we could consider making the commit request async, and consider a commit only completed after the commit response is returned. Of course it means more complicated logic.

Reply:
Thinking about this once more, we could actually keep the `commitOffsetNeeded` flag, and instead extract 
```
if (eosEnabled) {
    producer.commitTransaction();
    transactionInFlight = false;
    if (startNewTransaction) {
        producer.beginTransaction();
        transactionInFlight = true;
    }
}
```
If we put this after of the `if (commitOffsetNeeded) { }` block, we should be fine.
Note, that within the block, we still need:
```
if (eosEnabled) {
    producer.sendOffsetsToTransaction(consumedOffsetsAndMetadata, applicationId);
} else {
    consumer.commitSync(consumedOffsetsAndMetadata);
}
```

@guozhangwang WDYT about this?

Reply:
Even without EOS and assuming KIP-211 is in place, we'd probably still to commit regularly as a pre-requisite to fix https://issues.apache.org/jira/browse/KAFKA-6502.

Reply:
Fixing KAFKA-6502 might also not require to remove the flag -- keeping both separate makes sense to me.

@jadireddi Can you please address this comment and add a test. Thanks!

Reply:
Fixed.